### PR TITLE
Sharing map variable-height trees

### PR DESCRIPTION
**Pull request by danpoe**
This changes the sharing map such that not all leafs are stored at the same depth (which was depth 6 previously), but that the tree grows dynamically as elements are inserted.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by danpoe** [comment]
On a set of methods from Apache Tika, the sum of the runtimes of the sharing map methods (timing just those) improved by 22%. Overall symex was 4% faster.

File: src/util/sharing_map.h [hunk 1]
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
   iterate(n, f);
 }
 
+SHARING_MAPT(void)::add_item_if_not_shared(
```
Comment (smowton) [inline src/util/sharing_map.h:793]: Document this method

File: src/util/sharing_map.h [hunk 2]
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
...
+      }
+    }
+    else
+    {
+      SM_ASSERT(ip->is_container());
+
+      if(ip->shares_with(container))
+        return;
+
+      const leaft &l1 = container.get_container().front();
```
Comment (smowton) [inline src/util/sharing_map.h]: Suggest making `l`, `ll`, `l1` more descriptive, they're pretty easy to muddle as it stands

File: src/util/sharing_map.h [hunk 3]
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
...
+
+      const leaf_listt &ll = ip->get_container();
+      SM_ASSERT(!ll.empty());
+
+      for(const auto &l : ll)
```
Comment (smowton) [inline src/util/sharing_map.h]: Comment explaining this list? It looks like you're sure `container` doesn't have multiple entries but `ip->get_container()` might do, and if it does and an entry is shared that means all further entries are too?

File: src/util/sharing_map.h [hunk 4]
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
...
+      {
+        if(l1.shares_with(l))
+          return;
+
+        if(l1.get_key() == l.get_key())
```
Comment (smowton) [inline src/util/sharing_map.h]: Why don't we do anything when differing keys occur in `container` and `ip->get_container()`? Is that impossible for some reason?

File: src/util/sharing_map.h [hunk 5]
```diff
@@ -917,21 +995,28 @@ ::get_delta_view(
...
+      return ip;
 
     key >>= chunk;
   }
 
-  SM_ASSERT(ip->is_container());
-
-  return ip;
+  UNREACHABLE;
+  return nullptr;
```
Comment (smowton) [inline src/util/sharing_map.h:1025]: I think some compilers will complain about this unreachable statement

File: src/util/sharing_map.h [hunk 6]
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
   num--;
 }
 
+SHARING_MAPT2(, innert *)::migrate(
```
Comment (smowton) [inline src/util/sharing_map.h:1108]: Document this methid

File: unit/util/sharing_map.cpp [hunk 7]
```diff
@@ -126,6 +131,163 @@ void sharing_map_internals_test()
     sm.count_unmarked_nodes(false, marked, true);
     REQUIRE(marked.size() >= 2);
   }
+
+  SECTION("single map shape")
```
Comment (smowton) [inline unit/util/sharing_map.cpp:135]: Describe this in some more detail, what is shape?

File: unit/util/sharing_map.cpp [hunk 8]
```diff
@@ -214,15 +214,14 @@ void sharing_map_internals_test()
 
       REQUIRE(delta_view.size() == 2);
 
-      REQUIRE(!delta_view[0].in_both);
+      REQUIRE(!delta_view[0].is_in_both_maps());
```
Comment (smowton) [inline unit/util/sharing_map.cpp]: Squash this commit

File: src/util/sharing_map.h [hunk 9]
```diff
@@ -1283,7 +1288,7 @@ SHARING_MAPT2(optionalt<std::reference_wrapper<const, mapped_type>>)::find(
 
 SHARING_MAPT(const std::string)::not_found_msg="key not found";
 
-SHARING_MAPT(const std::size_t)::bits = 18;
+SHARING_MAPT(const std::size_t)::bits = 30;
```
Comment (smowton) [inline src/util/sharing_map.h:1295]: Surely belongs in its own commit

File: src/util/sharing_map.h [hunk 10]
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
...
+
+    if(ip->is_internal())
+    {
+      const to_mapt &m = ip->get_to_map();
+      SM_ASSERT(!m.empty());
+
+      for(const auto &item : m)
+      {
+        const innert *i = &item.second;
+        stack.push(i);
```
Comment (tautschnig) [inline src/util/sharing_map.h]: I don't think `i` adds any value here, I'd just do `stack.push(&item.second);`

File: src/util/sharing_map.h [hunk 11]
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
...
+
+      const leaft &l1 = container.get_container().front();
+
+      const leaf_listt &ll = ip->get_container();
+      SM_ASSERT(!ll.empty());
+
+      for(const auto &l : ll)
+      {
+        if(l1.shares_with(l))
+          return;
```
Comment (tautschnig) [inline src/util/sharing_map.h:838]: I would have expected `continue` - why is it ok to return?

File: src/util/sharing_map.h [hunk 12]
```diff
@@ -779,6 +791,73 @@ ::gather_all(const innert &n, delta_viewt &delta_view) const
...
+
+        if(l1.get_key() == l.get_key())
+        {
+          // element is in both maps and not shared
+          delta_view.push_back(
+            {true, l1.get_key(), l1.get_value(), l.get_value()});
+          return;
```
Comment (tautschnig) [inline src/util/sharing_map.h:843]: Again, I would have expected `continue` instead of returning.

File: src/util/sharing_map.h [hunk 13]
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
...
 }
 
+SHARING_MAPT2(, innert *)::migrate(
+  const std::size_t step,
+  const std::size_t key_suffix,
+  const std::size_t bit_last,
+  innert *ip)
+{
+  SM_ASSERT(step < steps - 1);
+  SM_ASSERT(ip != nullptr);
```
Comment (tautschnig) [inline src/util/sharing_map.h]: If it must not be a null pointer, why not pass in a reference?

File: src/util/sharing_map.h [hunk 14]
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
...
+
+  // Copy the container
+  innert container_copy(child);
+
+  // Delete existing container
+  ip->remove_child(bit_last);
+
+  // Add internal node
+  ip = ip->add_child(bit_last);
+  SM_ASSERT(ip->empty());
```
Comment (tautschnig) [inline src/util/sharing_map.h:1138]: I would just declare a local pointer variable instead of re-using the `ip` parameter.

File: src/util/sharing_map.h [hunk 15]
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
...
+
+  std::size_t i = step + 1;
+  std::size_t key = key_suffix;
+
+  key_existing >>= chunk;
+  key >>= chunk;
+
+  SM_ASSERT(i < steps);
+
+  do
```
Comment (tautschnig) [inline src/util/sharing_map.h:1150]: How about using a `for` loop instead: `for(std::size_t i = step + 1; i < steps; ++i)`?

File: src/util/sharing_map.h [hunk 16]
```diff
@@ -1001,13 +1098,138 @@ SHARING_MAPT(void)::erase(const key_type &k)
...
+
+    key >>= chunk;
+    key_existing >>= chunk;
+
+    i++;
+  } while(i < steps);
+
+  leaft leaf_copy(as_const(&container_copy)->get_container().front());
+  leaf_listt &c = ip->get_container();
+  c.push_front(leaf_copy);
```
Comment (tautschnig) [inline src/util/sharing_map.h]: `c` doesn't seem to add much value: `ip->get_container().push_front(leaf_copy);`

File: src/util/sharing_map.h [hunk 17]
```diff
@@ -922,6 +923,7 @@ ::get_delta_view(
         if(!child.shares_with(*ip2))
         {
           stack.push(stack_itemt(&child, ip2));
+          level_stack.push(0xff);
```
Comment (tautschnig) [inline src/util/sharing_map.h]: Maybe use `optionalt`, or else name that constant.

File: src/util/sharing_map.h (hunk 3 above)
Comment (danpoe) [reply src/util/sharing_map.h]: Yes, `add_item_if_not_shared()` is called by `get_delta_view()` only for containers which only have a single element. If we know that the leaf in the first container has the same key as a leaf in the second container (either because they share or the keys compare equal), then since the maps cannot contain duplicate keys, we can ignore all further keys in the second container since they will be different from the key of the leaf in the first container. We can ignore the further leafs in the second container since `map1.get_delta_view(map2, ...)` does not include items that are only in `map2`.

File: src/util/sharing_map.h (hunk 11 above)
Comment (danpoe) [reply src/util/sharing_map.h:838]: That's because if `l1` shares with `l`, then since the maps don't contain duplicate keys, all further keys in the container must be different from `l1.get_key()`. Since `map1.get_delta_view(map2, ...)` does not include items that are only in `map2`, we can return immediately.

File: src/util/sharing_map.h (hunk 12 above)
Comment (danpoe) [reply src/util/sharing_map.h:843]: Similar to above, since we have found a matching key, all other keys will be different and don't have to be included in the `delta_view`.

File: src/util/sharing_map.h (hunk 4 above)
Comment (danpoe) [reply src/util/sharing_map.h]: Ah yes there's a case missing for when we don't find a matching key. The bug would only appear when there are hash collisions. I'll fix it and add a test.

File: src/util/sharing_map.h (hunk 15 above)
Comment (danpoe) [reply src/util/sharing_map.h:1150]: For the first iteration we don't need to check the condition so a do-while is probably fine.

File: src/util/sharing_map.h (hunk 14 above)
Comment (danpoe) [reply src/util/sharing_map.h:1138]: Changed

File: src/util/sharing_map.h (hunk 9 above)
Comment (danpoe) [reply src/util/sharing_map.h:1295]: Done

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 38c8388639e406c7e79109f6731584872a870fec).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/111936752

**Review by tautschnig — APPROVED** [review]
